src/combinatorics/colex.lean

```diff
@@ -0,0 +1,292 @@
+  { simp only [mem_union, mem_sdiff] at hk,
+    cases hk,
+    { right,
+      right,
+      refine ⟨k, λ t th, _, hk.2, hk.1⟩,
+      specialize z t,
+      by_contra h₂,
+      simp only [mem_union, mem_sdiff, id.def] at z,
+      rw [not_iff, iff_iff_and_or_not_and_not, not_not, and_comm] at h₂,
+      apply not_le_of_lt th (z h₂) },
+    { left,
+      refine ⟨k, λ t th, _, hk.2, hk.1⟩,
+      specialize z t,
+      by_contra h₃,
+      simp only [mem_union, mem_sdiff, id.def] at z,
+      rw [not_iff, iff_iff_and_or_not_and_not, not_not, and_comm, or_comm] at h₃,
+      apply not_le_of_lt th (z h₃) }, },
+  rw nonempty_iff_ne_empty,
+  intro a,
+  simp only [union_eq_empty_iff, sdiff_eq_empty_iff_subset] at a,
+  apply h₁ (subset.antisymm a.1 a.2)
+end
+
+instance [linear_order α] : is_trichotomous (finset.colex α) (<) := ⟨lt_trichotomy⟩
+
+-- It should be possible to do this computably but it doesn't seem to make any difference for now.
+noncomputable instance [linear_order α] : linear_order (finset.colex α) :=
+{ le_refl := λ A, or.inr rfl,
+  le_trans := le_trans,
+  le_antisymm := λ A B AB BA, AB.elim (λ k, BA.elim (λ t, (asymm k t).elim) (λ t, t.symm)) id,
+  le_total := λ A B,
+          (lt_trichotomy A B).elim3 (or.inl ∘ or.inl) (or.inl ∘ or.inr) (or.inr ∘ or.inl),
+  decidable_le := classical.dec_rel _,
+  ..finset.colex.has_le }
+
+instance [linear_order α] : is_incomp_trans (finset.colex α) (<) :=
+begin
+  constructor,
+  rintros A B C ⟨nAB, nBA⟩ ⟨nBC, nCB⟩,
+  have: A = B := ((lt_trichotomy A B).resolve_left nAB).resolve_right nBA,
+  have: B = C := ((lt_trichotomy B C).resolve_left nBC).resolve_right nCB,
+  rw [‹A = B›, ‹B = C›, and_self],
+  apply irrefl
+end
+
+instance [linear_order α] : is_strict_weak_order (finset.colex α) (<) := {}
+instance [linear_order α] : is_strict_total_order (finset.colex α) (<) := {}
+
+/-- Colex is an extension of the base ordering on α. -/
+lemma singleton_lt_iff_lt [linear_order α] {r s : α} :
+  ({r} : finset α) <ᶜ {s} ↔ r < s :=
+begin
+  rw colex.lt_def,
+  simp only [mem_singleton, ← and_assoc, exists_eq_right],
+  split,
+  { rintro ⟨q, p⟩,
+    apply lt_of_le_of_ne _ (ne.symm p),
+    contrapose! p,
+    rw (q p).1 rfl },
+  { intro a,
+    exact ⟨λ z hz, iff_of_false (ne_of_gt (trans hz a)) (ne_of_gt hz), ne_of_gt a⟩ }
+end
+
+/--
+If A is before B in colex, and everything in B is small, then everything in A is small.
+-/
+lemma max_colex [linear_order α] {A B : finset α} (t : α) (h₁ : A <ᶜ B) (h₂ : ∀ x ∈ B, x < t) :
+  ∀ x ∈ A, x < t :=
+begin
+  rw colex.lt_def at h₁,
+  rcases h₁ with ⟨k, z, _, _⟩,
+  intros x hx,
+  apply lt_of_not_ge,
+  intro a,
+  refine not_lt_of_ge a (h₂ x _),
+  rwa ← z,
+  apply lt_of_lt_of_le (h₂ k ‹_›) a,
+end
+
+/-- If everything in A is less than k, we can bound the sum of powers. -/
+-- TODO: This might belong somewhere else.
+lemma binary_sum_nat {k : ℕ} {A : finset ℕ} (h₁ : ∀ {x}, x ∈ A → x < k) :
```

Comment:
This would probably be better outside the `colex` namespace at least, even if you leave it in this file